pecos/core/utils/clustering.hpp

```diff
@@ -77,6 +101,33 @@ struct Tree {
         seed_for_nodes.resize(nodes.size());
     }
 
+    struct ClusteringSampler {
+        // scheduler for sampling
+        ClusteringSamplerParam* param_ptr;
+        size_t warmup_layers;
+        size_t depth;
+
+        ClusteringSampler(ClusteringSamplerParam* param_ptr, size_t depth): param_ptr(param_ptr), depth(depth) {
+            warmup_layers = size_t(depth * param_ptr->warmup_layer_rate);
```

Comment:
What if user passing negative values for `warmup_layer_rate` in the `param_ptr`?